crypto/pkcs11key/pkcs11key.go

```diff
@@ -128,71 +204,67 @@ func New(module, slot, pin, privLabel string) (ps *PKCS11Key, err error) {
 		}
 	}
 	if !gotModulus || !gotExponent {
-		ps.Destroy()
-		return
+		return noKey, errors.New("public key missing either modulus or exponent")
 	}
-	ps.publicKey = rsa.PublicKey{
+	return rsa.PublicKey{
 		N: n,
 		E: e,
-	}
-
-	return
+	}, nil
 }
 
-// Destroy tears down a PKCS11Key.
-//
-// This method must be called before the PKCS11Key is GC'ed, in order
-// to ensure that the PKCS#11 module itself is properly finalized and
-// destroyed.
-//
-// The idiomatic way to do this (assuming no need for a long-lived
-// signer) is as follows:
-//
-//   ps, err := NewPKCS11Signer(...)
-//   if err != nil { ... }
-//   defer ps.Destroy()
+
+// Destroy tears down a PKCS11Key by closing the session. It should be
+// called before the key gets GC'ed, to avoid leaving dangling sessions.
 func (ps *PKCS11Key) Destroy() {
-	if ps.module != nil {
-		ps.module.Finalize()
-		ps.module.Destroy()
+	if ps.session != nil {
+		// NOTE: We do not want to call module.Logout here. module.Logout applies
+		// application-wide. So if there are multiple sessions active, the other ones
+		// would be logged out as well, causing CKR_OBJECT_HANDLE_INVALID next
+		// time they try to sign something. It's also unnecessary to log out explicitly:
+		// module.CloseSession will log out once the last session in the application is
+		// closed.
+		ps.sessionMu.Lock()
+		ps.module.CloseSession(*ps.session)
+		ps.session = nil
+		ps.sessionMu.Unlock()
 	}
 }
 
-func (ps *PKCS11Key) openSession() (session pkcs11.SessionHandle, err error) {
-	// Find slot by description
+func (ps *PKCS11Key) openSession() (pkcs11.SessionHandle, error) {
+	var noSession pkcs11.SessionHandle
 	slots, err := ps.module.GetSlotList(true)
 	if err != nil {
-		return
+		return noSession, err
 	}
+
 	for _, slot := range slots {
-		slotInfo, err := ps.module.GetSlotInfo(slot)
+		// Check that token label matches.
+		tokenInfo, err := ps.module.GetTokenInfo(slot)
 		if err != nil {
+			return noSession, err
+		}
+		if tokenInfo.Label != ps.tokenLabel {
 			continue
 		}
 
-		if slotInfo.SlotDescription == ps.slotDescription {
-			// Open session
-			session, err = ps.module.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION)
-			if err != nil {
-				return session, err
-			}
-
-			// Login
-			if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
-				return session, err
-			}
-
+		// Open session
+		session, err := ps.module.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION)
+		if err != nil {
 			return session, err
 		}
-	}
 
-	err = errors.New("slot not found")
-	return
-}
+		// Login
+		// Note: Logged-in status is application-wide, not per session. But in
+		// practice it appears to be okay to login to a token multiple times with the same
+		// credentials.
+		if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
```

Comment:
I've done this same thing before with other PKCS11 modules and it's worked fine.